src/event_generator.hpp

```diff
@@ -0,0 +1,209 @@
+};
+
+// Generator that for events in a generic sequence
+template <typename Seq>
+struct seq_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+    seq_generator(Seq& events):
+        events_(events),
+        it_(std::begin(events_))
+    {
+        EXPECTS(std::is_sorted(events_.begin(), events_.end()));
+    }
+
+    postsynaptic_spike_event next() override {
+        return it_==events_.end()? terminal_pse(): *it_;
+    }
+
+    void pop() override {
+        if (it_!=events_.end()) {
+            ++it_;
+        }
+    }
+
+    void reset() override {
+        it_ = events_.begin();
+    }
+
+    void advance(time_type t) override {
+        it_ = std::lower_bound(events_.begin(), events_.end(), t, event_time_less());
+    }
+
+private:
+
+    const Seq& events_;
+    typename Seq::const_iterator it_;
+};
+
+// Generates a stream of events
+//  * with the first event at time t_start
+//  * subsequent events at t_start+n*dt, ∀ n ∈ [0, ∞)
+//  * with a set target and weight
+struct regular_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+
+    regular_generator(time_type tstart, time_type dt, cell_member_type target, float weight):
+        t_start_(tstart), step_(0), dt_(dt),
+        target_(target), weight_(weight)
+    {}
+
+    postsynaptic_spike_event next() override {
+        return {target_, t_start_+(step_*dt_), weight_};
+    }
+
+    void pop() override {
+        ++step_;
+    }
+
+    void advance(time_type t0) override {
+        t0 = t0<t_start_? t_start_: t0;
+        step_ = (t0-t_start_)/dt_;
+        // Finding the smallest value for step_ that satisfies the condition
+        // that time() >= t0 is unfortunately a horror show because floating
+        // point precission.
+        while (step_ && time()>=t0) {
+            --step_;
+        }
+        while (time()<t0) {
+            ++step_;
+        }
+    }
+
+    void reset() override {
+        step_ = 0;
+    }
+
+private:
+    time_type time() const {
+        return t_start_ + step_*dt_;
+    }
+
+    time_type t_start_;
+    std::size_t step_;
+    time_type dt_;
+    cell_member_type target_;
+    float weight_;
+};
+
+// Generates a stream of events
+//  * At times described by a Poisson point process with rate 1/dt
+template <typename RandomNumberEngine>
+struct poisson_generator: public event_generator {
+    using pse = postsynaptic_spike_event;
+
+    poisson_generator(time_type tstart,
+                      time_type dt,
+                      cell_member_type target,
+                      float weight,
+                      RandomNumberEngine rng):
+        t_start_(tstart),
+        exp_(1./dt),
+        reset_state_(std::move(rng)),
+        target_(target),
+        weight_(weight)
+    {
+        reset();
+    }
+
+    postsynaptic_spike_event next() override {
+        return {target_, next_, weight_};
+    }
+
+    void pop() override {
+        next_ += exp_(rng_);
```

Comment:
As we have discussed, this might not work with the counter based random number generators as Random123. 

However, I think the easiest way to make this work with Random123 (or some other counter-based generator) is to make a wrapper around Random123 which will be compatible with the standard library c++ distribution so that we can invoke it as exp_(random123_generator).